tmt/steps/__init__.py

```diff
@@ -392,12 +404,136 @@ def go(self):
+            return phases
+
+        # Process provided options
+        for option in options:
+            # Parse the step:phase format
+            matched = re.match(r'(\w+)(:(\w+))?', option)
+            if matched:
+                step_name, _, phase = matched.groups()
+            if not matched or step_name not in STEPS:
+                raise tmt.utils.GeneralError(f"Invalid step '{option}'.")
+            # Check phase format, convert into int, use end by default
+            try:
+                phase = int(phase)
+            except TypeError:
+                phase = PHASE_END
+            except ValueError:
+                # Convert 'start' and 'end' aliases
+                try:
+                    phase = dict(start=PHASE_START, end=PHASE_END)[phase]
+                except KeyError:
+                    raise tmt.utils.GeneralError(f"Invalid phase '{phase}'.")
+            # Store the phase for given step
+            try:
+                phases[step_name].append(phase)
+            except KeyError:
+                phases[step_name] = [phase]
+        return phases
+
+
+class Reboot(ActionStep):
+    """ Reboot guest """
+
+    # Which kind of reboot
+    _hard = False
+
+    # True if reboot enabled
+    _enabled = False
+
+    # Guest name provided via CLI
+    _guest_name = None
+
+    def __init__(self, step, order):
+        """ Initialize relations, store the reboot order """
+        super().__init__(parent=step, name='reboot')
+        self.order = order
+
+    @classmethod
+    def command(cls, method_class=None, usage=None):
+        """ Create the reboot command """
+        @click.command()
+        @click.pass_context
+        @click.option(
+            '-s', '--step', metavar='STEP[:PHASE]', multiple=True,
+            help='Log in during given phase of selected step(s).')
+        @click.option(
+            '--hard', metavar='REBOOT_HARD', is_flag=True,
+            help='Hard reboot of provisioned machine.')
+        @click.option(
+            '--name', metavar='MACHINE_NAME', multiple=False,
+            default=None, help='Provisioned machine name.')
+        def reboot(context, **kwargs):
+            """
+            Reboot machine.
+            """
+            Reboot._save_context(context)
+            Reboot._enabled = True
+
+        return reboot
+
+    @classmethod
+    def plugins(cls, step):
+        """ Return list of reboot instances for given step """
+        if not Reboot._enabled:
+            return []
+        return [Reboot(step, phase) for phase in cls.phases(step)]
+
+    def go(self, *args, **kwargs):
+        """ Reboot the guest(s) """
+
+        self._guest_name = self.opt("name")
+        # Run the interactive command
+        self.info('reboot', 'Rebooting machine', color='yellow')
+        for guest in self.parent.plan.provision.guests():
+            # Attempt to push the workdir to the guest
+            try:
+                guest.push()
```

Comment:
Why do we push the workdir to guest here?

Reply:
I too was a bit surprised by this. I then thought about it a bit more and doesn't it make sense to sync the workdir with the guest before rebooting (to have all up-to-date data once the machine is up)? Not sure if such case even exists, just a food for thought

Reply:
I see, this possibly might be useful/needed in the future. For now, if we are not sure, I'd vote for keeping it simple, removing the `push()` and adding only when/if there is a clear need for it.

Reply:
Removed. Solved. I wasn't sure in which state we want to have machine before reboot.